api/pkg/s3/objectget.go

```diff
@@ -24,65 +33,152 @@ import (
+			if err == ErrorInvalidRange {
+				WriteErrorResponse(response, request, s3error.ErrInvalidRange)
+				return
+			}
+			// log the error.
+			log.Errorln("invalid request range, err:", err)
+		}
 	}
-	if client == nil {
-		response.WriteError(http.StatusInternalServerError, NoSuchBackend.Error())
+
+	// Validate pre-conditions if any.
+	if err = checkPreconditions(request.Request.Header, object); err != nil {
+		// set object-related metadata headers
+		response.AddHeader("Last-Modified", time.Unix(object.LastModified, 0).UTC().Format(http.TimeFormat))
+
+		if object.Etag != "" {
+			response.ResponseWriter.Header()["ETag"] = []string{"\"" + object.Etag + "\""}
+		}
+		if err == s3error.ContentNotModified { // write only header if is a 304
+			WriteErrorResponseHeaders(response, err)
+		} else {
+			WriteErrorResponse(response, request, err)
+		}
 		return
 	}
-	log.Infof("enter the download method")
-	body, s3err := client.GET(&object, ctx, int64(start), int64(end))
-	log.Infof("out  the download method")
-	if s3err != NoError {
-		response.WriteError(http.StatusInternalServerError, s3err.Error())
+
+	// Get the object.
+	startOffset := int64(0)
+	length := object.Size
+	if hrange != nil {
+		startOffset = hrange.OffsetBegin
+		length = hrange.GetLength()
+	}
+	stream, err := s.s3Client.GetObject(ctx, &pb.GetObjectInput{Bucket:bucketName, Key:objectKey, Offset: startOffset, Length: length,})
+	if err != nil {
+		log.Errorln("get object failed, err:%v", err)
+		WriteErrorResponse(response, request, err)
 		return
 	}
-	buf := new(bytes.Buffer)
-	buf.ReadFrom(body)
-	response.Write(buf.Bytes())
-	*/
-	log.Infof("Init multipart upload[bucketName=%s, objectKey=%s] successfully.\n", bucketName, objectKey)
+	defer stream.Close()
+
+	// Indicates if any data was written to the http.ResponseWriter
+	dataWritten := false
+	// io.Writer type which keeps track if any data was written.
+	writer := funcToWriter(func(p []byte) (int, error) {
+		if !dataWritten {
+			// Set headers on the first write.
+			// Set standard object headers.
+			SetObjectHeaders(response, object, hrange)
+
+			// Set any additional requested response headers.
+			setGetRespHeaders(response.ResponseWriter, request.Request.URL.Query())
+			dataWritten = true
+		}
+		n, err := response.Write(p)
+		if n > 0 {
+
+		}
+		return n, err
+	})
+
+	s3err := int32(0)
+	eof := false
+	left := length
+	for !eof && left > 0 {
+		rsp, err := stream.Recv()
+		if err != nil && err != io.EOF {
+			log.Errorln("recv err", err)
+			break
+		}
+		if err == io.EOF {
+			eof = true
+		}
+
+		if rsp.ErrorCode != int32(s3error.ErrNoErr) {
+			s3err = rsp.ErrorCode
+			log.Errorf("received s3 service error, error code:", rsp.ErrorCode)
+			break
+		}
+		if len(rsp.Data) == 0 {
+			continue
+		}
+		_, err = writer.Write(rsp.Data)
+		if err != nil {
+			log.Errorln("failed to write data to client. err:", err)
+			break
+		}
+		left -= int64(len(rsp.Data))
+	}
+
+	if !dataWritten {
```

Comment:
dataWritten will always be 'true' because it was set on line 136.

Reply:
if file size is zero, dataWritten is false